vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,859 @@
+                }
+
+                if vring
+                    .add_used(desc_chain.head_index(), event.as_slice().len() as u32)
+                    .is_err()
+                {
+                    log::error!("Couldn't write event data to the ring");
+                    return Err(VuInputError::UnexpectedWriteVringError);
+                }
+            } else {
+                // Now cannot get available descriptor, which means the host cannot process
+                // event data in time and overrun happens in the backend. In this case,
+                // we simply drop the incomping input event and notify guest for handling
+                // events. At the end, it returns Ok(false) so can avoid exiting the thread loop.
+                self.ev_list.clear();
+
+                vring
+                    .signal_used_queue()
+                    .map_err(|_| VuInputError::SendNotificationFailed)?;
+
+                return Ok(false);
+            }
+        }
+
+        self.ev_list.drain(0..last_sync_index + 1);
+        Ok(true)
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_queue(&mut self, vring: &VringRwLock) -> Result<bool> {
+        let events = match self.ev_dev.fetch_events() {
+            Err(err) => {
+                log::error!("Error::DescriptorWriteFailed: {}", err);
+                return Err(VuInputError::UnexpectedFetchEventError);
+            }
+            Ok(e) => e,
+        };
+
+        for event in events {
+            let ev_raw_data = VuInputEvent {
+                ev_type: event.event_type().0,
+                code: event.code(),
+                value: event.value() as u32,
+            };
+            self.ev_list.push_back(ev_raw_data);
+        }
+
+        self.process_event(vring)?;
+
+        vring
+            .signal_used_queue()
+            .map_err(|_| VuInputError::SendNotificationFailed)?;
+
+        Ok(true)
+    }
+
+    pub fn read_event_config(&self) -> Result<VuInputConfig> {
+        let mut counter: u8 = 0;
+        let mut index: u8 = 0;
+        let mut cfg: [u8; VIRTIO_INPUT_CFG_SIZE] = [0; VIRTIO_INPUT_CFG_SIZE];
+
+        let func: unsafe fn(nix::libc::c_int, &mut [u8]) -> nix::Result<libc::c_int> =
+            match self.subsel {
+                EV_KEY => eviocgbit_key,
+                EV_ABS => eviocgbit_absolute,
+                EV_REL => eviocgbit_relative,
+                EV_MSC => eviocgbit_misc,
+                EV_SW => eviocgbit_switch,
+                _ => {
+                    return Err(VuInputError::HandleEventUnknownEvent);
+                }
+            };
+
+        // SAFETY: Safe as the file is a valid event device, the kernel will only
+        // update the correct amount of memory in func.
+        if unsafe { func(self.ev_dev.get_raw_fd(), &mut cfg) }.is_err() {
+            return Err(VuInputError::UnexpectedInputDeviceError);
+        }
+
+        for val in &cfg {
```

Comment:
index should probably defined right before this. Same for counter

Reply:
I would like to change code with enumerate():

```
        let mut size: u8 = 0;
        for (index, val) in cfg.iter().enumerate() {
            if *val != 0 {
                size = (index + 1) as u8;
            }
        }
```